### PR TITLE
Logcollector integration tests T0: age option

### LINKED ISSUE #1241: Logcollector integration tests T0: Check age option
It is necessary to add tests for the logcollector [age](https://documentation.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html?highlight=localfile#age) option as part of https://github.com/wazuh/wazuh-qa/issues/1027

# Objectives
- [x] Test age option works correctly, ignoring files that have not been modified since specified values.
- [x] Test after change host date, logcollector still works and age value is applied correctly

# Organization

## Tasks
- [x] Recopilate logcollector logs that allow us to check if a file is being monitored or ignored.
- [x] Create a new test (tier 0) to check age option works correctly.
- [x] Create a new test (tier 0) to check logcollector works after changing host date.
- [x] Testing and debug new tests.
- [ ] Documentation new tests and  functions.

### PR DESCRIPTION
**Pull request by Rebits**
|Related issue|
|---|
|Closes #1241 |

## Description

This PR adds logcollector tests for [age](https://documentation-dev.wazuh.com/current/user-manual/reference/ossec-conf/localfile.html?highlight=only%20future%20event#age) option. 
  - [x] **Test age basic**: Check that logcollector ignores files that have not been modified for longer than the time specified by age. Also, check that logcollector monitor the files when they are modified.
  - [x] **Test age datetime changed**: Check that logcollector age option works correctly when the date of the system changed.

These tests are valid for manager and agent in Linux and Windows system.

### Linux - Manager

test_manager

### Linux - Agent

test_agent

### Windows - Agent

test_windows_age

### Index

INDEX

### Test age basic

BASIC

### Test age DateTime changed

datetime

## Update 07/05/2021

Due to the age option bug, described in this issue https://github.com/wazuh/wazuh/issues/8438#issuecomment-834246543, `test_age_basic` has been marked as xfail. 

- **Build**: https://ci.wazuh.info/job/Test_integration/2971/console

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -417,14 +419,32 @@ def configure_local_internal_options(get_local_internal_options):
...
         control_service('restart')
     else:
         yield
 
 
+@pytest.fixture(scope='function')
+def create_file_structure(get_files_list):
+    """
+
+    """
```
Comment (BraulioV) [inline tests/integration/conftest.py:433]: Docstrings are missing.

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -417,14 +419,32 @@ def configure_local_internal_options(get_local_internal_options):
...
+    for file in get_files_list:
+        os.makedirs(file['folder_path'], exist_ok=True, mode=0o777)
+        f = open(f"{file['folder_path']}{file['filename']}", "w+").close()
```
Comment (BraulioV) [inline tests/integration/conftest.py:436]:
```suggestion
        f = open(f"{file['folder_path']}{file['filename']}", "w").close()
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+from wazuh_testing.tools.time import time_to_seconds
+
+# Marks
+pytestmark = pytest.mark.tier(level=0)
+
+# Configuration
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+configurations_path = os.path.join(test_data_path, 'wazuh_age.yaml')
+
+WINDOWS_FOLDER_PATH = r'C:\testing_age' + '\\'
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:22]: I recommend you to change this directory to `%temppdir`. This is the temporary directory for Windows. You can use the function `gettempdir()` to detect this directory independently of the OS.

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    {'location': f'{folder_path}*', 'log_format': 'syslog', 'age': '4000s'},
+    {'location': f'{folder_path}*', 'log_format': 'syslog', 'age': '5m'},
+    {'location': f'{folder_path}*', 'log_format': 'syslog', 'age': '500m'},
+    {'location': f'{folder_path}*', 'log_format': 'syslog', 'age': '9h'},
+    {'location': f'{folder_path}*', 'log_format': 'syslog', 'age': '200d'},
+]
+
+configurations = load_wazuh_configurations(configurations_path, __name__,
+                                           params=parameters,
+                                           metadata=metadata)
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:84]:
```suggestion
configurations = load_wazuh_configurations(configurations_path, __name__, params=parameters, metadata=metadata)
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                                                                  f"{file['folder_path']}{file['filename']}",
+                                                                  prefix=prefix)
+        wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                error_message=f'{file["filename"]} was not detected')
+
+        if int(age_seconds) <= int(file['age']):
+            log_callback = logcollector.callback_ignoring_file(
+                f"{file['folder_path']}{file['filename']}", prefix=prefix)
+            wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                    error_message=f'{file["filename"]} was not ignored')
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:133]:
```suggestion
            log_callback = logcollector.callback_ignoring_file(f"{file['folder_path']}{file['filename']}", prefix=prefix)
            wazuh_log_monitor.start(timeout=5, callback=log_callback, error_message=f'{file["filename"]} was not ignored')
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                log_callback = logcollector.callback_ignoring_file(
+                    f"{file['folder_path']}{file['filename']}", prefix=prefix)
+                wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                        error_message=f'{file["filename"]} was not ignored')
+
+    for file in file_structure:
+
+        f = open(f"{file['folder_path']}{file['filename']}", "a")
+        f.write(file['content'])
+        f.close()
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:146]:
```suggestion
       with open(os.path.join(file['folder_path'], file['filename']), 'a') as file:
           file.write(file['content'])
```

File: tests/integration/test_logcollector/test_age/test_age_datetime_changed.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# Marks
+pytestmark = pytest.mark.tier(level=0)
+
+# Configuration
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+configurations_path = os.path.join(test_data_path, 'wazuh_age.yaml')
+
+WINDOWS_FOLDER_PATH = 'C:\\testing_age' + '\\'
+LINUX_FOLDER_PATH = '/tmp/testing_age/'
+DAEMON_NAME = "wazuh-logcollector"
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_datetime_changed.py:27]: Same as before.

**Review by BraulioV — CHANGES_REQUESTED** [review]

File: tests/integration/conftest.py (hunk 1 above)
Comment (Rebits) [reply tests/integration/conftest.py]: Done 41fd7460dc64d554987256b75777f84190bbdb4c

File: tests/integration/test_logcollector/test_age/test_age_basic.py (hunk 3 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_basic.py]: Done 41fd7460dc64d554987256b75777f84190bbdb4c and a6da7de49152ea31d69d3cf25dcb3b82e288dc95

File: tests/integration/test_logcollector/test_age/test_age_basic.py (hunk 4 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_basic.py]: Done 41fd7460dc64d554987256b75777f84190bbdb4c

File: tests/integration/test_logcollector/test_age/test_age_basic.py (hunk 5 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_basic.py]: Done 41fd7460dc64d554987256b75777f84190bbdb4c

File: tests/integration/test_logcollector/test_age/test_age_basic.py (hunk 6 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_basic.py]: Done 41fd7460dc64d554987256b75777f84190bbdb4c

File: tests/integration/test_logcollector/test_age/test_age_datetime_changed.py (hunk 7 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_datetime_changed.py]: Done a6da7de49152ea31d69d3cf25dcb3b82e288dc95

File: deps/wazuh_testing/wazuh_testing/logcollector.py [hunk 8]
```diff
@@ -169,6 +169,24 @@ def callback_invalid_location_pattern(location, prefix=monitoring.LOG_COLLECTOR_
...
+
+
+def callback_reading_syslog_message(message, prefix=monitoring.LOG_COLLECTOR_DETECTOR_PREFIX):
+    msg = fr"DEBUG: Reading syslog message: '{message}'"
+    return monitoring.make_callback(pattern=msg, prefix=prefix, escape=True)
+
+
+def callback_read_line_from_file(n_lines, filename, prefix=monitoring.LOG_COLLECTOR_DETECTOR_PREFIX):
+    msg = fr"DEBUG: Read {n_lines} lines from {filename}"
+    return monitoring.make_callback(pattern=msg, prefix=prefix, escape=True)
```
Comment (BraulioV) [inline deps/wazuh_testing/wazuh_testing/logcollector.py:189]: Docstrings are missing. Also, keep in mind following the PEP-8.

File: tests/integration/conftest.py [hunk 9]
```diff
@@ -379,7 +381,7 @@ def close_sockets(receiver_sockets):
     """Close the sockets connection gracefully."""
     for socket in receiver_sockets:
         try:
-            # We flush the buffer before closing connection if connection is TCP
+            # We flush the buffer before closing connection if connection is TCPt_files_list:
```
Comment (BraulioV) [inline tests/integration/conftest.py:384]:
```suggestion
            # We flush the buffer before closing the connection if the protocol is TCP:
```

File: tests/integration/conftest.py [hunk 10]
```diff
@@ -417,14 +419,33 @@ def configure_local_internal_options(get_local_internal_options):
...
         conf.set_wazuh_local_internal_options(backup_options_lines)
 
         control_service('restart')
     else:
         yield
 
 
+@pytest.fixture(scope='function')
+def create_file_structure(get_files_list):
+    """ Creates the directory structure specified for a test case"""
```
Comment (BraulioV) [inline tests/integration/conftest.py:431]:
```suggestion
    """Creates the directory structure specified for a test case."""
```

The `Args` section is missing

File: tests/integration/conftest.py [hunk 11]
```diff
@@ -417,14 +419,33 @@ def configure_local_internal_options(get_local_internal_options):
...
+    for file in get_files_list:
+        absolute_file_path = os.path.join(file['folder_path'], file['filename'])
+        os.makedirs(file['folder_path'], exist_ok=True, mode=0o777)
+        os.path.join
```
Comment (BraulioV) [inline tests/integration/conftest.py:435]:
```suggestion
```

File: tests/integration/conftest.py [hunk 12]
```diff
@@ -417,14 +419,33 @@ def configure_local_internal_options(get_local_internal_options):
...
+        open(absolute_file_path, "w").close()
+
+        if 'age' in file:
+            fileinfo = os.stat(absolute_file_path)
+            os.utime(absolute_file_path, (fileinfo.st_atime - file['age'],
+                                                                  fileinfo.st_mtime - file['age']))
```
Comment (BraulioV) [inline tests/integration/conftest.py:441]:
```suggestion
            os.utime(absolute_file_path, (fileinfo.st_atime - file['age'], fileinfo.st_mtime - file['age']))
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+folder_path = os.path.join(tempfile.gettempdir(), 'wazuh_testing_age')
+folder_path_regex = os.path.join(folder_path, '*')
+
+local_internal_options = {'logcollector.vcheck_files': 0}
+
+if sys.platform == 'win32':
+    prefix = AGENT_DETECTOR_PREFIX
+else:
+    prefix = LOG_COLLECTOR_DETECTOR_PREFIX
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:31]: This was changed in the master branch. You will need to updated with the latest changes.

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        absolute_file_path = os.path.join(file['folder_path'], file['filename'])
+        wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
+
+        log_callback = logcollector.callback_file_matches_pattern(cfg['location'], absolute_file_path, prefix=prefix)
+        wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                error_message=f'{file["filename"]} was not detected')
+
+        if int(age_seconds) <= int(file['age']):
+            log_callback = logcollector.callback_ignoring_file(
+                absolute_file_path, prefix=prefix)
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:130]:
```suggestion
            log_callback = logcollector.callback_ignoring_file(absolute_file_path, prefix=prefix)
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+            wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                    error_message=f'{file["filename"]} was not ignored')
+
+        else:
+            with pytest.raises(TimeoutError):
+                log_callback = logcollector.callback_ignoring_file(absolute_file_path,
+                                                                   prefix=prefix)
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:137]:
```suggestion
                log_callback = logcollector.callback_ignoring_file(absolute_file_path, prefix=prefix)
```

File: tests/integration/test_logcollector/test_age/test_age_basic.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+                wazuh_log_monitor.start(timeout=5, callback=log_callback,
+                                        error_message=f'{file["filename"]} was not ignored')
+
+    for file in file_structure:
+        with open(absolute_file_path, 'a') as file_to_write:
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_basic.py:142]: I think this variable should be redefined. Other wise, the value of `absolute_file_path` will be the one from the last iteration of the previous loop.

File: tests/integration/test_logcollector/test_age/test_age_datetime_changed.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+
+now_date = datetime.now()
+folder_path = os.path.join(tempfile.gettempdir(), 'wazuh_testing_age')
+folder_path_regex = os.path.join(folder_path, '*')
+
+if sys.platform == 'win32':
+    prefix = AGENT_DETECTOR_PREFIX
+else:
+    prefix = LOG_COLLECTOR_DETECTOR_PREFIX
```
Comment (BraulioV) [inline tests/integration/test_logcollector/test_age/test_age_datetime_changed.py:35]: Same as before.

**Review by BraulioV — CHANGES_REQUESTED** [review]

File: deps/wazuh_testing/wazuh_testing/logcollector.py (hunk 8 above)
Comment (Rebits) [reply deps/wazuh_testing/wazuh_testing/logcollector.py]: Done 0f925459f2174454140dcf7aedc8ae418930e12a

File: tests/integration/conftest.py (hunk 9 above)
Comment (Rebits) [reply tests/integration/conftest.py]: Done 0f925459f2174454140dcf7aedc8ae418930e12a

File: tests/integration/conftest.py (hunk 10 above)
Comment (Rebits) [reply tests/integration/conftest.py]: Done 5a5fabfccc0faab98a1ed77559881d0bebd9a227

File: tests/integration/conftest.py (hunk 11 above)
Comment (Rebits) [reply tests/integration/conftest.py]: Done 0f925459f2174454140dcf7aedc8ae418930e12a

File: tests/integration/test_logcollector/test_age/test_age_basic.py (hunk 14 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_basic.py]: Done 0f925459f2174454140dcf7aedc8ae418930e12a

File: tests/integration/test_logcollector/test_age/test_age_datetime_changed.py (hunk 17 above)
Comment (Rebits) [reply tests/integration/test_logcollector/test_age/test_age_datetime_changed.py]: Done 0f925459f2174454140dcf7aedc8ae418930e12a

**Review by BraulioV — APPROVED** [review]

**Pull request merged**